File: Easy/1189_Maximum_Number_of_Balloons/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int maxNumberOfBalloons(string text) {
        unordered_map<char, int> freq;

        for (char c : text) {
            freq[c]++;
        }

        return min(freq['b'], min(freq['a'],  min(freq['l'] / 2, min(freq['o'] / 2, freq['n']))));
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    string text1 = "nlaebolko";
    cout << solution.maxNumberOfBalloons(text1) << endl;

    // test cases 2
    string text2 = "loonbalxballpoon";
    cout << solution.maxNumberOfBalloons(text2) << endl;

    // test cases 3
    string text3 = "leetcode";
    cout << solution.maxNumberOfBalloons(text3) << endl;

    return 0;
}

File: Easy/1189_Maximum_Number_of_Balloons/Solution.java



class Solution {
    public int maxNumberOfBalloons(String text) {
        int[] freq = new int[26];

        for (char c : text.toCharArray()) {
            freq[c - 'a']++;
        }

        return Math.min(
            Math.min(freq['b' - 'a'], freq['a' - 'a']),
            Math.min(
                freq['l' - 'a'] / 2,
                Math.min(freq['o' - 'a'] / 2, freq['n' - 'a'])
            )
        );
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String text1 = "nlaebolko";
        System.out.println(solution.maxNumberOfBalloons(text1));

        // test cases 2
        String text2 = "loonbalxballpoon";
        System.out.println(solution.maxNumberOfBalloons(text2));

        // test cases 3
        String text3 = "leetcode";
        System.out.println(solution.maxNumberOfBalloons(text3));
        
    }
}